File: leetcode/0005-longest-palindromic-substring.cc
// https://leetcode.com/problems/longest-palindromic-substring/
// Medium

#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    string longestPalindrome(string s) {
        int size = s.length(), maxSoFar = 0, maxLength = 0;
        vector<vector<bool>> v (size, vector<bool> (size, false));
        for (int i = 0; i < size; ++i) {
            v[i][i] = true;
        }
        for (int i = 1; i < size; ++i) {
            v[i][i-1] = true;
        }
        for (int j = 1; j < size; ++j) {
            for (int i = 0; i + j < size; ++i) {
                if (s[i] == s[i+j] && v[i+1][i+j-1]) {
                    v[i][i+j] = true;
                    if (maxLength < j) {
                        maxSoFar = i;
                        maxLength = j;
                    }
                }
            }
        }
        return s.substr(maxSoFar, maxLength + 1);
    }
};

File: leetcode/python/0005-longest-palindromic-substring.py
# https://leetcode.com/problems/longest-palindromic-substring/
# Medium

class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[0]*len(s) for _ in range(len(s))]
        for i in range(len(s)):
            dp[i][i] = True
        
        maxLen = 1
        ret = s[0]
        for i in range(1, len(s)):
            for j in range(len(s)-i):
                k = j+i
                if s[j] == s[k] and (dp[j+1][k-1] or k-j == 1):
                    dp[j][k] = True
                    if i+1 > maxLen:
                        maxLen = i+1
                        ret = s[j:k+1]
        return ret
